File: backend/src/agent/task_catalog.py
"""任务分类与列表展示辅助。"""

from __future__ import annotations

_CATEGORY_LABELS = {
    "travel": "差旅办事",
    "meeting": "会议预约",
    "workpackage": "工时填报",
    "email": "邮件撰写",
    "other": "综合办事",
}

_TOOL_TAGS: dict[str, str] = {
    "travel_apply": "差旅申请",
    "flight_book": "机票预订",
    "hotel_book": "酒店预订",
    "email_notify": "邮件通知",
    "meeting_book": "会议预约",
    "room_book": "会议室",
    "workpackage_fill": "工时填报",
}


def infer_category(steps: list[dict]) -> str:
    tools = {s.get("tool") for s in steps if s.get("tool")}
    if "workpackage_fill" in tools:
        return "workpackage"
    if "meeting_book" in tools or "room_book" in tools:
        return "meeting"
    if "travel_apply" in tools or "flight_book" in tools or "hotel_book" in tools:
        return "travel"
    if "email_notify" in tools:
        return "email"
    return "other"


def infer_tags(steps: list[dict]) -> list[str]:
    tags: list[str] = []
    for step in steps:
        tool = step.get("tool")
        if tool in ("user_confirm",):
            continue
        label = _TOOL_TAGS.get(tool) or step.get("action")
        if label and label not in tags:
            tags.append(label)
    return tags


def count_completed_steps(steps: list[dict]) -> int:
    return sum(1 for s in steps if s.get("status") == "completed")


def effective_status(raw_status: str, steps: list[dict]) -> str:
    if raw_status == "cancelled":
        return "cancelled"
    if raw_status == "completed":
        return "completed"
    if raw_status == "failed":
        return "failed"
    if steps and all(s.get("status") == "completed" for s in steps):
        return "completed"
    if raw_status in ("running", "pending"):
        return "running"
    return raw_status or "running"


def category_label(category: str) -> str:
    return _CATEGORY_LABELS.get(category, "综合办事")


def build_task_summary(record) -> dict:
    steps = list(record.steps_json or [])
    total = record.total_steps or len(steps) or 0
    completed = count_completed_steps(steps)
    category = infer_category(steps)
    status = effective_status(record.status, steps)
    progress = int(completed / total * 100) if total else 0

    return {
        "id": record.id,
        "session_id": record.session_id,
        "goal": record.goal,
        "status": status,
        "raw_status": record.status,
        "category": category,
        "category_label": category_label(category),
        "tags": infer_tags(steps),
        "completed_steps": completed,
        "total_steps": total,
        "progress_percent": progress,
        "current_step": record.current_step,
        "replan_count": record.replan_count,
        "created_at": record.created_at.isoformat() if record.created_at else "",
    }

File: frontend/src/utils/taskCatalog.ts
import type { Task, TaskStep } from '@/types'
import type { TaskCategory, TaskDisplayStatus, TaskSummary } from '@/types'

const CATEGORY_LABELS: Record<TaskCategory, string> = {
  travel: '差旅办事',
  meeting: '会议预约',
  gn_meeting: '国能会议',
  workpackage: '工时填报',
  leave: '请假申请',
  email: '邮件撰写',
  info_collect: '信息收集',
  other: '综合办事',
}

const TOOL_TAGS: Record<string, string> = {
  travel_apply: '差旅申请',
  flight_book: '机票预订',
  hotel_book: '酒店预订',
  email_notify: '邮件通知',
  meeting_book: '会议预约',
  gn_meeting_book: '国能会议',
  room_book: '会议室',
  workpackage_fill: '工时填报',
  leave_apply: '请假申请',
  info_collect_publish: '信息收集',
}

export function inferTaskCategory(steps: TaskStep[]): TaskCategory {
  const tools = new Set(steps.map((s) => s.tool))
  if (tools.has('info_collect_publish')) return 'info_collect'
  if (tools.has('leave_apply')) return 'leave'
  if (tools.has('workpackage_fill')) return 'workpackage'
  if (tools.has('gn_meeting_book')) return 'gn_meeting'
  if (tools.has('meeting_book') || tools.has('room_book')) return 'meeting'
  if (tools.has('travel_apply') || tools.has('flight_book') || tools.has('hotel_book')) {
    return 'travel'
  }
  if (tools.has('email_notify')) return 'email'
  return 'other'
}

export function inferTaskTags(steps: TaskStep[]): string[] {
  const tags: string[] = []
  for (const step of steps) {
    if (step.tool === 'user_confirm') continue
    const label = TOOL_TAGS[step.tool] ?? step.action
    if (label && !tags.includes(label)) tags.push(label)
  }
  return tags
}

export function countCompletedSteps(steps: TaskStep[]): number {
  return steps.filter((s) => s.status === 'completed').length
}

export function effectiveTaskStatus(task: Task): TaskDisplayStatus {
  if (task.status === 'cancelled') return 'cancelled'
  if (task.status === 'completed') return 'completed'
  if (task.status === 'failed') return 'failed'
  if (task.steps.length && task.steps.every((s) => s.status === 'completed')) {
    return 'completed'
  }
  return 'running'
}

export function categoryLabel(category: TaskCategory): string {
  return CATEGORY_LABELS[category] ?? '综合办事'
}

export function buildTaskSummary(task: Task): TaskSummary {
  const completed = countCompletedSteps(task.steps)
  const total = task.total_steps || task.steps.length
  const category = inferTaskCategory(task.steps)
  const status = effectiveTaskStatus(task)
  return {
    id: task.id,
    session_id: task.session_id,
    goal: task.goal,
    status,
    raw_status: task.status,
    category,
    category_label: categoryLabel(category),
    tags: inferTaskTags(task.steps),
    completed_steps: completed,
    total_steps: total,
    progress_percent: total ? Math.round((completed / total) * 100) : 0,
    current_step: task.current_step,
    replan_count: task.replan_count,
    created_at: task.created_at,
  }
}

export const CATEGORY_ICONS: Record<TaskCategory, string> = {
  travel: '✈️',
  meeting: '📅',
  gn_meeting: '🎥',
  workpackage: '📊',
  leave: '🏖️',
  email: '✉️',
  info_collect: '📝',
  other: '📋',
}

export const STATUS_LABELS: Record<TaskDisplayStatus, string> = {
  running: '进行中',
  completed: '已完成',
  cancelled: '已取消',
  failed: '失败',
}
